### PR TITLE
Acl integration tests

### PR DESCRIPTION
**Pull request by AtapinDmitry**
ACL integration tests added

### CODE REVIEW

**Comment by pfi79** [comment]
@zemtsov please see the pr

File: internal/config/config.go [hunk 1]
```diff
@@ -60,13 +62,29 @@ func SetConfig(stub shim.ChaincodeStubInterface) error {
...
 		if validator.GetPublicKey() == "" {
-			cfg.Validators[i].KeyType = helpers.DefaultPublicKeyType()
+			return fmt.Errorf(ErrValidatorEmpty, i)
 		}
 		// gost key can't be used as a validator's key
 		if !helpers.ValidatePublicKeyType(validator.GetKeyType(), pb.KeyType_gost.String()) {
-			return fmt.Errorf("invalid key type: %s", validator.GetPublicKey())
+			return fmt.Errorf(ErrValidatorInvalidKeyType, i, validator.GetPublicKey())
 		}
 	}
```
Comment (pfi79) [inline internal/config/config.go:89]: Where does the validator replay check take place?

File: tests/integration/integration_test.go [hunk 2]
```diff
@@ -1,4 +1,4 @@
-package integrational
+package integration
 
 import (
```
Comment (pfi79) [inline tests/integration/integration_test.go:3]: Why was this file left behind?

File: tests/integration/tools.go [hunk 3]
```diff
@@ -0,0 +1,21 @@
...
+	_ "github.com/anoideaopen/foundation/test/chaincode/fiat"
+	_ "github.com/anoideaopen/foundation/test/chaincode/industrial"
+	_ "github.com/anoideaopen/robot"
+	_ "github.com/hyperledger/fabric/cmd/configtxgen"
+	_ "github.com/hyperledger/fabric/cmd/cryptogen"
+	_ "github.com/hyperledger/fabric/cmd/discover"
+	_ "github.com/hyperledger/fabric/cmd/orderer"
+	_ "github.com/hyperledger/fabric/cmd/osnadmin"
+	_ "github.com/hyperledger/fabric/cmd/peer"
+	_ "github.com/hyperledger/fabric/integration/chaincode/simple/cmd"
```
Comment (pfi79) [inline tests/integration/tools.go:20]: I guess not all imports are needed here

File: tests/integration/cmn/client/acl.go [hunk 4]
```diff
@@ -0,0 +1,119 @@
...
+)
+
+const MultisignKeyDelimeter = "/"
+
+func CheckUserMultisigned(network *nwo.Network, peer *nwo.Peer, user *UserFoundationMultisigned) {
+	Eventually(func() string {
+		sess, err := network.PeerUserSession(peer, "User1", commands.ChaincodeQuery{
+			ChannelID: cmn.ChannelAcl,
+			Name:      cmn.ChannelAcl,
+			Ctor:      cmn.CtorFromSlice([]string{"checkKeys", user.PublicKey()}),
```
Comment (zemtsov) [inline tests/integration/cmn/client/acl.go:28]: a package tests/common defines constants for ACL functions. "checkKeys" might be common.FnCheckKeys

File: tests/integration/cmn/client/acl.go [hunk 5]
```diff
@@ -0,0 +1,119 @@
...
+
+		out := sess.Out.Contents()[:len(sess.Out.Contents())-1] // skip line feed
+		resp := &pb.AclResponse{}
+		err = proto.Unmarshal(out, resp)
+		Expect(err).NotTo(HaveOccurred())
+		if err != nil {
+			return fmt.Sprintf("failed to unmarshal response: %v", err)
+		}
+
+		addr := base58.CheckEncode(resp.GetAddress().GetAddress().GetAddress()[1:], resp.GetAddress().GetAddress().GetAddress()[0])
```
Comment (zemtsov) [inline tests/integration/cmn/client/acl.go:44]: I'd introduce a variable addressBytes := resp.GetAddress().GetAddress().GetAddress() to make this line a little shorter

File: tests/integration/cmn/client/acl.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+			return fmt.Sprintf("Error: expected %s, received %s", user.AddressBase58Check, addr)
+		}
+
+		return ""
+	}, network.EventuallyTimeout, time.Second).Should(BeEmpty())
+}
+
+// AddMultisig adds multisigned user
+func AddMultisig(network *nwo.Network, peer *nwo.Peer, orderer *nwo.Orderer, n int, user *UserFoundationMultisigned) {
+	ctorArgs := []string{"addMultisig", strconv.Itoa(n), client.NewNonceByTime().Get()}
```
Comment (zemtsov) [inline tests/integration/cmn/client/acl.go:55]: again, a function name can be taken from the constant of common package

File: tests/integration/cmn/client/acl.go [hunk 7]
```diff
@@ -0,0 +1,119 @@
...
+	})
+	Expect(err).NotTo(HaveOccurred())
+	Eventually(sess, network.EventuallyTimeout).Should(gexec.Exit(0))
+	Expect(sess.Err).To(gbytes.Say("Chaincode invoke successful. result: status:200"))
+
+	CheckUserMultisigned(network, peer, user)
+}
+
+// ChangeMultisigPublicKey changes public key for multisigned user by validators
+func ChangeMultisigPublicKey(network *nwo.Network, peer *nwo.Peer, orderer *nwo.Orderer,
```
Comment (zemtsov) [inline tests/integration/cmn/client/acl.go:81]: very long list of arguments. I'd reformat it placing one arg per line. like this:
func foo(
   arg1 string,
   arg2 int,
   ...
) error {}

File: tests/integration/cmn/client/user.go [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+// signMessage - sign arguments with private key in ed25519
+func signMessage(privateKey ed25519.PrivateKey, msgToSign []byte) []byte {
+	sig := ed25519.Sign(privateKey, msgToSign)
+	return sig
+}
+
+// verifyEd25519 - verify publicKey with message and signed message
+func verifyEd25519(publicKey ed25519.PublicKey, bytesToSign []byte, sMsg []byte) error {
+	if !ed25519.Verify(publicKey, bytesToSign, sMsg) {
+		err := errors.New("valid signature rejected")
```
Comment (zemtsov) [inline tests/integration/cmn/client/user.go:88]: variable err looks redundant here

File: cc/chaincode.go [hunk 9]
```diff
@@ -87,3 +75,29 @@ func (c *ACL) Invoke(stub shim.ChaincodeStubInterface) peer.Response {
 
 	return ccInvoke(stub, args)
 }
+
+// readConfig reads & validates ACL config
+func (c *ACL) readConfig(stub shim.ChaincodeStubInterface) error {
```
[the remark this answers is not in the document]
Comment (pfi79) [reply cc/chaincode.go:80]: rename function readAndValidateConfig

**Review by pfi79 — APPROVED** [review]

**Pull request merged**